### PR TITLE
gh-125118: don't copy arbitrary values to _Bool in the struct module

### PR DESCRIPTION
**Pull request by skirpichev**
memcopy'ing arbitrary values to _Bool variable triggers undefined behaviour, e.g.:
```c
/* a.c */
#include <stdbool.h>
#include <stdio.h>
#include <string.h>
int main(void)
{
    _Bool x;
    char y = 2;
    memcpy(&x, &y, 1);
    printf("%d\n", x != 0);
    return 0;
}
```
```
$ gcc-12 a.c -Wall && ./a.out
2
$ clang-15 a.c -Wall && ./a.out
0
$ gcc-12 a.c -Wall -fsanitize=undefined && ./a.out
a.c:9:5: runtime error: load of value 2, which is not a valid value for type '_Bool'
0
$ clang-15 a.c -Wall -fsanitize=undefined && ./a.out
a.c:9:20: runtime error: load of value 2, which is not a valid value for type '_Bool'
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior a.c:9:20 in
0
```

Cridits to Alex Gaynor.

* Issue: gh-125118

### CODE REVIEW

**Review by vstinner — APPROVED** [review]
LGTM, but I would prefer to have a second review on this change.

@encukou: You worked on a similar issue some time ago, would you mind to review this change?

**Comment by encukou** [comment]
Yes, please wait for my review. I'll get to it this week.

**Review by ZeroIntensity — APPROVED** [review]
I think this looks fine. Out of curiosity, what bug was this causing in practice?

**Comment by alex** [comment]
It was triggering UBSAN warnings, because it was UB. AFAIK it wasn't misbehaving on any platform _yet_, but it'd have been within the compiler's right to rewrite `!= 0` as `== 1`, and that would have changed behavior.

Should we add a `static_assert(sizeof(_Bool) == 1);`? Otherwise LGTM.

**Comment by skirpichev** [comment]
@ZeroIntensity, you could try added tests on gcc-12 (PASS) and clang-17 (FAIL).  I suspect specific versions aren't important.

File: Modules/_struct.c
Comment (colesbury) [inline Modules/_struct.c]: If we want to handle the cases where `_Bool` is not one byte, I think we can do something like the following:

```c
_Bool zero = 0;
return PyBool_FromLong(memcmp(p, &zero, sizeof(_Bool)) != 0);
```

Otherwise, we probably want the `static_assert` like @alex suggested.

**Comment by encukou** [comment]
As far as I can see, the safest assumption we can make is that `false` is all zero bytes.

To allow bigger `sizeof(_Bool)`, that would be:
```c
    static const _Bool false_bool = 0;
    return PyBool_FromLong(memcmp(p, &false_bool, sizeof(_Bool)));
```

File: Modules/_struct.c
Comment (encukou) [reply Modules/_struct.c]: I swear I didn't see this before I posted my version! :)

**Comment by vstinner** [comment]
> static const _Bool false_bool = 0;

I don't think that static is needed here.

File: Modules/_struct.c
Comment (vstinner) [inline Modules/_struct.c]: Call it "b" or another name different than the parameter "f" :-)

File: Modules/_struct.c
Comment (vstinner) [inline Modules/_struct.c]: You forgot one closing parenthesis, no?

**Review by alex — APPROVED** [review]

File: Modules/_struct.c
Comment (vstinner) [inline Modules/_struct.c]:
```suggestion
    const _Bool bool_false = false;
    return PyBool_FromLong(memcmp(p, &bool_false, sizeof(_Bool)));
```

Comment (vstinner) [reply Modules/_struct.c]: I'm afraid that names starting with underscore can conflict with an existing name.

**Review by vstinner — APPROVED** [review]
LGTM.

File: Modules/_struct.c
Comment (encukou) [inline Modules/_struct.c]:
```suggestion
    const _Bool bool_false = 0;
```

Or use and `false` and `bool` (not `_Bool`) , but then include `<stdbool.h>` directly, don't rely on something else pulling it in. These will become C keywords in C23 but we aren't there yet.

Comment (skirpichev) [reply Modules/_struct.c]: Ah, I thought it's included in Python.h like stdint.h.

**Comment by encukou** [comment]
Thank you!

**Comment by skirpichev** [comment]
Working on backports...

**Pull request merged**
